src/cluster_legacy.c

```diff
@@ -2996,14 +2996,20 @@ int clusterProcessPacket(clusterLink *link) {
         /* Add this node if it is new for us and the msg type is MEET.
          * In this stage we don't try to add the node with the right
          * flags, replicaof pointer, and so forth, as this details will be
-         * resolved when we'll receive PONGs from the node. */
+         * resolved when we'll receive PONGs from the node. The exception
+         * to this is the flag that indicates extensions are supported, as
+         * we want to send extensions right away in the return PONG in order
+         * to reduce the amount of time needed to stabilize the shard ID. */
         if (!sender && type == CLUSTERMSG_TYPE_MEET) {
             clusterNode *node;
 
             node = createClusterNode(NULL, CLUSTER_NODE_HANDSHAKE);
             serverAssert(nodeIp2String(node->ip, link, hdr->myip) == C_OK);
             getClientPortFromClusterMsg(hdr, &node->tls_port, &node->tcp_port);
             node->cport = ntohs(hdr->cport);
+            if (hdr->mflags[0] & CLUSTERMSG_FLAG0_EXT_DATA) {
+                node->flags |= CLUSTER_NODE_EXTENSIONS_SUPPORTED;
+            }
```

Comment:
Move it around here? line 2936
```
    if (sender && (hdr->mflags[0] & CLUSTERMSG_FLAG0_EXT_DATA)) {
        sender->flags |= CLUSTER_NODE_EXTENSIONS_SUPPORTED;
    }
```

but i am ok with it. Not a blocker

Reply:
In this circumstance, I believe we need to add it here so that the newly created node has this flag immediately in the PONG writing call on line 3017 

```
if (!sender && type == CLUSTERMSG_TYPE_MEET) clusterProcessGossipSection(hdr, link);
        /* Anyway reply with a PONG */
        clusterSendPing(link, CLUSTERMSG_TYPE_PONG);
    }
```

This allows this node to be marked as extensions-supported right away in the logic on line 3814

```
if (link->node && nodeSupportsExtensions(link->node))
```

Reply:
I made a mistake and need to update the commit message - this does save time to stabilization however not in the way I had anticipated - the node does not attach to the link here so does not attach extensions to the PONG, but instead saving the flag here saves one hop, it seems by allowing extensions to be attached to the next message instead of waiting for a second PING/PONG

Reply:
It is worth exploring if there would be negative issues with attaching the node to the link in this meet, I believe that would have the intended effect

Reply:
>It may be worth exploring if there would be negative issues with attaching the node to the link in this meet, I believe that would have the intended effect

I also feel that we should fix the link here but in addition, I think we need another change to create the node with its real name on [L3006](https://github.com/valkey-io/valkey/pull/778/files#diff-2515500619600c5a1e7a8d9aaa8761a6071314cd226c3b7ee4d0e054691f0a8eR3006), i.e.,

```c
    node = createClusterNode(hdr->sender, CLUSTER_NODE_HANDSHAKE);
```

Would you like to take a shot at it, @bentotten?

Reply:
Accidentally made a review comment instead of continuing the thread: https://github.com/valkey-io/valkey/pull/778#pullrequestreview-2189498791

Adding the name here unfortunately causes problems with handshake resolution on the next ping